### ISSUE TITLE
Progressive shutdown - shedding

### ISSUE THREAD

**Issue by Jmgr** [liftbridge-io/liftbridge#317]
Currently, when a Liftbridge server shuts down it stops being a leader for its partitions. If many partitions exist that will result in a flurry of Raft events. Would it be possible to trigger a progressive shutdown to prevent this? Have you had some thought about this @tylertreat?

**Comment by tylertreat** [comment]
Yes, this is something I've thought a bit about, especially as it relates to rolling cluster upgrades. I think a graceful shutdown would make sense. There would be a few components to this:

1. If the server is leader for any partitions, transfer leadership to another replica (invoke a `ChangeLeaderOp` in Raft) and remove self from ISR (`ShrinkISROp`). This should be down gradually to avoid a flood of Raft ops. Also interrupt any clients currently subscribed.
2. If server is follower for any partitions, remove self from ISR (`ShrinkISROp`). This should be done gradually to avoid a flood of Raft ops. Also interrupt any clients currently subscribed.
3. At this point, probably reject any client requests, e.g. publish or subscribe.
4. If the server shutting down is the metadata leader, transfer leadership to another node. Perform a Raft barrier to ensure all preceding Raft ops have been applied.
5. Remove self from Raft group. Need to think through how this works when rejoining, e.g. in the case of restarting/upgrading a node.
6. Shut down the server.

**Comment by xe-nvdk** [comment]
Hi @Jmgr, Thank you for this detailed feature request. The project has recently been revived under new maintainership ([Basekick Labs](https://github.com/basekick-labs)), and we're reviewing the backlog. 

Current Status: Partially implemented. The infrastructure for progressive shutdown exists:

✅ ChangeLeaderOp and ShrinkISROp are implemented
✅ Individual partition leaders have graceful shutdown with WaitGroup coordination
❌ Server-level Stop() does NOT use these for progressive shedding
❌ No gradual ISR removal for followers
❌ No client request rejection during shutdown phase
❌ No metadata leadership transfer

The core issue remains: when Server.Stop() is called, all partitions lose leadership simultaneously, causing the Raft event flood you described. We've added this to our roadmap. The implementation would follow @tylertreat's original outline - gradually transfer leadership, shed ISR membership, reject new clients, transfer metadata leadership, then shut down. This is a significant feature that would benefit rolling upgrades and cluster maintenance. 

We'll update this issue when we have a timeline.
